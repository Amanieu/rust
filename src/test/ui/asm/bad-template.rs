// only-x86_64

#![feature(asm)]

fn main() {
    let mut foo = 0;
    unsafe {
        asm!("{}");
        //~^ ERROR invalid reference to argument at index 0
        asm!("{1}", in(reg) foo);
        //~^ ERROR invalid reference to argument at index 1
        //~^^ WARN asm argument not used in template
        asm!("{a}");
        //~^ ERROR there is no argument named `a`
        asm!("{}", a = in(reg) foo);
        //~^ ERROR invalid reference to argument at index 0
        //~^^ WARN asm argument not used in template
        asm!("{1}", a = in(reg) foo);
        //~^ ERROR invalid reference to argument at index 1
        //~^^ WARN asm argument not used in template
        asm!("{}", in("eax") foo);
        //~^ ERROR invalid reference to argument at index 0
        asm!("{:foo}", in(reg) foo);
        //~^ ERROR asm template modifier must be a single character
        asm!("", in(reg) 0, in(reg) 1);
        //~^ WARN asm arguments not used in template
    }
}
