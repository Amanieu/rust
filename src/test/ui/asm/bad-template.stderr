error: invalid reference to argument at index 0
  --> $DIR/bad-template.rs:8:15
   |
LL |         asm!("{}");
   |               ^^ from here
   |
   = note: no arguments were given

error: invalid reference to argument at index 1
  --> $DIR/bad-template.rs:10:15
   |
LL |         asm!("{1}", in(reg) foo);
   |               ^^^ from here
   |
   = note: there is 1 argument

error: there is no argument named `a`
  --> $DIR/bad-template.rs:13:15
   |
LL |         asm!("{a}");
   |               ^^^

error: invalid reference to argument at index 0
  --> $DIR/bad-template.rs:15:15
   |
LL |         asm!("{}", a = in(reg) foo);
   |               ^^   --------------- named argument
   |               |
   |               from here
   |
   = note: no positional arguments were given
note: named arguments cannot be referenced by position
  --> $DIR/bad-template.rs:15:20
   |
LL |         asm!("{}", a = in(reg) foo);
   |                    ^^^^^^^^^^^^^^^

error: invalid reference to argument at index 1
  --> $DIR/bad-template.rs:18:15
   |
LL |         asm!("{1}", a = in(reg) foo);
   |               ^^^ from here
   |
   = note: no positional arguments were given

error: invalid reference to argument at index 0
  --> $DIR/bad-template.rs:21:15
   |
LL |         asm!("{}", in("eax") foo);
   |               ^^   ------------- explicit register argument
   |               |
   |               from here
   |
   = note: no positional arguments were given
note: explicit register arguments cannot be used in the asm template
  --> $DIR/bad-template.rs:21:20
   |
LL |         asm!("{}", in("eax") foo);
   |                    ^^^^^^^^^^^^^

error: asm template modifier must be a single character
  --> $DIR/bad-template.rs:23:17
   |
LL |         asm!("{:foo}", in(reg) foo);
   |                 ^^^

warning: asm argument not used in template
  --> $DIR/bad-template.rs:10:21
   |
LL |         asm!("{1}", in(reg) foo);
   |                     ^^^^^^^^^^^
   |
   = note: `#[warn(unused_asm_arguments)]` on by default

warning: asm argument not used in template
  --> $DIR/bad-template.rs:15:20
   |
LL |         asm!("{}", a = in(reg) foo);
   |                    ^^^^^^^^^^^^^^^

warning: asm argument not used in template
  --> $DIR/bad-template.rs:18:21
   |
LL |         asm!("{1}", a = in(reg) foo);
   |                     ^^^^^^^^^^^^^^^

warning: asm arguments not used in template
  --> $DIR/bad-template.rs:25:18
   |
LL |         asm!("", in(reg) 0, in(reg) 1);
   |                  ^^^^^^^^^  ^^^^^^^^^

warning: asm arguments not used in template
  --> $DIR/bad-template.rs:28:18
   |
LL |         asm!("", in(reg) 0, in(reg) 1);
   |                  ^^^^^^^^^  ^^^^^^^^^

error: aborting due to 7 previous errors; 5 warnings emitted

